File: linkedlist/147.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

#不新建任何node, 完全复用原始节点
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head:
            return None
        
        cur = head.next #遍历链表的游标, 从第二个开始
        head.next = None #头节点作为新的起点
        while cur:
            dummy = head #重置dummy到头节点
            if cur.val <= dummy.val: #处理插入节点在头节点之前的情况
                temp = cur
                cur = cur.next
                temp.next = dummy
                head = temp
            else:
                while dummy.next and cur.val > dummy.next.val:
                    dummy = dummy.next
                if dummy.next is None: #处理cur是当前最大节点的情况
                    temp = cur
                    cur = cur.next
                    dummy.next = temp
                    temp.next = None
                else:                  #将cur插入dummy, dummy.next之间
                    temp = cur
                    cur = cur.next
                    temp.next = dummy.next
                    dummy.next = temp
                    
        return head
                
            
            

File: linkedlist/147.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertionSortList(head *ListNode) *ListNode {
    if head == nil {
        return head
    }   
    var cur, dummy, temp *ListNode
    
    cur = head.Next //遍历的起点
    head.Next = nil //排序后链表的头
    
    for cur != nil {
        dummy = head //重制dummy到头节点
        if cur.Val <= dummy.Val {
            //移动cur
            temp = cur
            cur = cur.Next
            //插入到head之前
            temp.Next = head
            head = temp
        }else {
            //找到cur在结果集里的位置
            for dummy.Next != nil && cur.Val > dummy.Next.Val {
                dummy = dummy.Next
            }
            //如果是cur是已知最大的, 放到最后
            if dummy.Next == nil {
                temp = cur
                cur = cur.Next
                temp.Next = nil
                dummy.Next = temp
            }else {
                //放在dummy后面
                temp = cur
                cur = cur.Next
                temp.Next = dummy.Next
                dummy.Next = temp
            }
        }
    }
    return head
}
